txhelpers/subsidy.go

```diff
@@ -68,17 +101,20 @@ func UltimateSubsidy(params *chaincfg.Params) int64 {
 	}
 
 	// Update the ultimate subsidy store.
-	ultimateSubsidies.Store(params, totalSubsidy)
+	ultimateSubsidies.store(nh, totalSubsidy)
 
 	return totalSubsidy
 }
 
 // RewardsAtBlock computes the PoW, PoS (per vote), and project fund subsidies
-// at for the specified block index, assuming a certain number of votes.
-func RewardsAtBlock(blockIdx int64, votes uint16, p *chaincfg.Params) (work, stake, tax int64) {
+// at for the specified block index, assuming a certain number of votes. The
+// stake reward is for a single vote. The total reward for the block is thus
+// work + stake * votes + tax.
+func RewardsAtBlock(blockIdx int64, votes uint16, p *chaincfg.Params, useDCP0010 bool) (work, stake, tax int64) {
 	subsidyCache := standalone.NewSubsidyCache(p)
-	work = subsidyCache.CalcWorkSubsidy(blockIdx, votes)
-	stake = subsidyCache.CalcStakeVoteSubsidy(blockIdx)
-	tax = subsidyCache.CalcTreasurySubsidy(blockIdx, votes, false) //!!!! what?
+	work = subsidyCache.CalcWorkSubsidyV2(blockIdx, votes, useDCP0010)
```

Comment:
That occurred to me while working on this update.  I think this was initially just laziness, but now I'm not clear what happens if you call, say `CalcWorkSubsidyV2` with the same `height` and `voters` twice, but with different `useDCP0010` values.  What gets cached?

Reply:
The full block subsidy, which is the same regardless of DCP0010, is what gets cached.  The result will be correct if you call it back to back with different `useDCP0010` values.

Reply:
Same if you change voters at the same height?  e.g. doing projections or hypothetical computations

I'm just trying to understand if there can be a package-level store for these `SubsidyCache`s per network.

Reply:
Correct, you can change all parameters to `CalcWorkSubsidyV2` and `CalcStakeVoteSubsidyV2` without any issues in terms of caching.   I was careful to update it in such a way that calling those back to back with different values won't break it.

The code in question (for work) is [here](https://github.com/decred/dcrd/blob/2e8abf6bf3840cfd6fa54b3c5a814678a3f1a508/blockchain/standalone/subsidy.go#L241-L329).

Notice that `calcWorkSubsidy` is just called with the different proportions depending on the flag and the result is calculated by applying the proportions to the full block subsidy (which only takes the height as a parameter).

It is only the full block subsidy that is cached, because that's the expensive part which involves iteratively applying reductions.

Reply:
OK, great.  Thanks for clarifying that.  I'll definitely hang on to each `subsidyCache` instance to reuse them.